[JSC] Concurrently computing CodeBlockHash

Constellation opened the pull request:
#### 82de5bb8778460c45a1eda6245f1de2b93f46952
<pre>
[JSC] Concurrently computing CodeBlockHash
<a href="https://bugs.webkit.org/show_bug.cgi?id=295000">https://bugs.webkit.org/show_bug.cgi?id=295000</a>
<a href="https://rdar.apple.com/154307835">rdar://154307835</a>

Reviewed by Yijia Huang.

This patch fixes an issue for concurrent CodeBlockHash. We put some
locks around CachedScriptSourceProvider / ScriptBufferSourceProvider
to ensure that we can touch these source code safely (do not ref them.
This is concurrently touched).

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::hash const):
* Source/JavaScriptCore/bytecode/CodeBlockHash.cpp:
(JSC::CodeBlockHash::CodeBlockHash):
* Source/JavaScriptCore/bytecode/CodeBlockHash.h:
* Source/JavaScriptCore/parser/SourceProvider.cpp:
(JSC::SourceProvider::codeBlockHashConcurrently):
* Source/JavaScriptCore/parser/SourceProvider.h:
(JSC::SourceProviderBufferGuard::provider):
* Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
* Source/WebCore/bindings/js/ScriptBufferSourceProvider.h:
* Source/WebCore/loader/TextResourceDecoder.cpp:
(WebCore::TextResourceDecoder::TextResourceDecoder):
(WebCore::TextResourceDecoder::create):
* Source/WebCore/loader/TextResourceDecoder.h:
(WebCore::TextResourceDecoder::contentType const):
(WebCore::TextResourceDecoder::usesEncodingDetector const):
* Source/WebCore/loader/cache/CachedScript.cpp:
(WebCore::CachedScript::script):
(WebCore::CachedScript::codeBlockHashConcurrently):
(WebCore::CachedScript::destroyDecodedData):
(WebCore::CachedScript::setBodyDataFrom):
* Source/WebCore/loader/cache/CachedScript.h:

Canonical link: <a href="https://commits.webkit.org/296668@main">https://commits.webkit.org/296668@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a8d18c26f110d30816838778626d35fb1c90143

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109245 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28903 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19331 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114453 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59516 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111208 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29585 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37492 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83026 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112193 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed webkitperl tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23547 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98397 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63476 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22936 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16542 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59078 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101761 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92914 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16582 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117569 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107803 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36289 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26852 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92038 "Found 1 new test failure: svg/text/timeout-long-text-content.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36661 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94660 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91847 "Found 2 new API test failures: /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed, /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36769 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14522 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32122 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17624 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36185 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41673 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/132072 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35871 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35792 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39203 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37562 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
```diff
@@ -68,6 +61,7 @@ class CachedScriptSourceProvider final : public JSC::SourceProvider, public Cach
 
     CachedResourceHandle<CachedScript> m_cachedScript;
     RefPtr<FragmentedSharedBuffer> m_buffer;
```
Do we need to remove the `m_buffer` in `CachedScriptSourceProvider` as well?

Constellation replied:
Right!

hyjorc1 commented on Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
```diff
@@ -68,6 +61,7 @@ class CachedScriptSourceProvider final : public JSC::SourceProvider, public Cach
 
     CachedResourceHandle<CachedScript> m_cachedScript;
     RefPtr<FragmentedSharedBuffer> m_buffer;
+    mutable Lock m_lock;
```
And we don't need this lock right? Since `CachedScript` has its own lock.

Constellation replied:
Right! Removed.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Concurrently computing CodeBlockHash
https://bugs.webkit.org/show_bug.cgi?id=295000
rdar://154307835

Reviewed by Yijia Huang.

This patch fixes an issue for concurrent CodeBlockHash. We put some
locks around CachedScriptSourceProvider / ScriptBufferSourceProvider
to ensure that we can touch these source code safely (do not ref them.
This is concurrently touched).

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::hash const):
* Source/JavaScriptCore/bytecode/CodeBlockHash.cpp:
(JSC::CodeBlockHash::CodeBlockHash):
* Source/JavaScriptCore/bytecode/CodeBlockHash.h:
* Source/JavaScriptCore/parser/SourceProvider.cpp:
(JSC::SourceProvider::codeBlockHashConcurrently):
* Source/JavaScriptCore/parser/SourceProvider.h:
(JSC::SourceProviderBufferGuard::provider):
* Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
* Source/WebCore/bindings/js/ScriptBufferSourceProvider.h:
* Source/WebCore/loader/TextResourceDecoder.cpp:
(WebCore::TextResourceDecoder::TextResourceDecoder):
(WebCore::TextResourceDecoder::create):
* Source/WebCore/loader/TextResourceDecoder.h:
(WebCore::TextResourceDecoder::contentType const):
(WebCore::TextResourceDecoder::usesEncodingDetector const):
* Source/WebCore/loader/cache/CachedScript.cpp:
(WebCore::CachedScript::script):
(WebCore::CachedScript::codeBlockHashConcurrently):
(WebCore::CachedScript::destroyDecodedData):
(WebCore::CachedScript::setBodyDataFrom):
* Source/WebCore/loader/cache/CachedScript.h:

Canonical link: https://commits.webkit.org/296668@main
